test/framework/src/main/java/org/elasticsearch/search/ErrorTraceHelper.java

```diff
@@ -26,16 +27,20 @@ public enum ErrorTraceHelper {
 
     public static BooleanSupplier setupErrorTraceListener(InternalTestCluster internalCluster) {
         final AtomicBoolean transportMessageHasStackTrace = new AtomicBoolean(false);
-        internalCluster.getDataNodeInstances(TransportService.class).forEach(ts -> ts.addMessageListener(new TransportMessageListener() {
-            @Override
-            public void onResponseSent(long requestId, String action, Exception error) {
-                TransportMessageListener.super.onResponseSent(requestId, action, error);
-                if (action.startsWith("indices:data/read/search")) {
-                    Optional<Throwable> throwable = ExceptionsHelper.unwrapCausesAndSuppressed(error, t -> t.getStackTrace().length > 0);
-                    transportMessageHasStackTrace.set(throwable.isPresent());
+        internalCluster.getDataNodeInstances(TransportService.class)
+            .forEach(ts -> ((MockTransportService) ts).addMessageListener(new TransportMessageListener() {
```

Comment:
Suggest `asInstanceOf(MockTransportService.class, ts)`, we sometimes catch a `ClassCastException` without failing the test.

Reply:
++